### PR TITLE
Clippy fixes

### PR DESCRIPTION
**Pull request by kwf**
This PR is a series of fixes to lints detected by clippy; all lints are now resolved. Clippy uncovered one performance optimization: a location where bounds checks could be elided in a loop by using an iterator.

### CODE REVIEW

File: crypto/src/action.rs [hunk 1]
```diff
@@ -36,13 +36,9 @@ impl TryFrom<transaction::Action> for Action {
         }
 
         match proto.action.unwrap() {
-            transaction::action::Action::Output(inner) => {
```
Comment (kwf) [inline crypto/src/action.rs:39]: Unnecessary `return`

File: crypto/src/action.rs [hunk 2]
```diff
@@ -36,13 +36,9 @@ impl TryFrom<transaction::Action> for Action {
...
-                return Ok(Action::Output(inner.try_into()?));
-            }
+            transaction::action::Action::Output(inner) => Ok(Action::Output(inner.try_into()?)),
 
-            transaction::action::Action::Spend(inner) => {
```
Comment (kwf) [inline crypto/src/action.rs:43]: Unnecessary `return`

File: crypto/src/action/output.rs [hunk 3]
```diff
@@ -71,7 +71,7 @@ impl Body {
         let note_commitment = note.commit();
 
         let ephemeral_key = esk.diversified_public(&note.diversified_generator());
-        let encrypted_note = note.encrypt(&esk);
```
Comment (kwf) [inline crypto/src/action/output.rs:74]: `esk` is already a reference, the `&` is dereferenced by the compiler

File: crypto/src/action/spend.rs [hunk 4]
```diff
@@ -110,9 +111,9 @@ impl Body {
     }
 }
 
-impl Into<Vec<u8>> for Body {
```
Comment (kwf) [inline crypto/src/action/spend.rs:113]: Replaced `Into` with `From`

File: crypto/src/address.rs [hunk 5]
```diff
@@ -3,7 +3,6 @@ use std::io::{Cursor, Read, Write};
 use anyhow::anyhow;
 use ark_serialize::CanonicalDeserialize;
 use bech32::{FromBase32, ToBase32, Variant};
-use decaf377;
```
Comment (kwf) [inline crypto/src/address.rs:6]: Unused root import

File: crypto/src/address.rs [hunk 6]
```diff
@@ -98,7 +97,7 @@ impl std::str::FromStr for Address {
     type Err = anyhow::Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let (hrp, data, variant) = bech32::decode(&s).unwrap();
```
Comment (kwf) [inline crypto/src/address.rs:101]: `s` is already a reference, Rust immediately dereferences the `&`

File: crypto/src/memo.rs [hunk 7]
```diff
@@ -34,7 +34,7 @@ impl MemoPlaintext {
     pub fn encrypt(&self, esk: &ka::Secret, address: &Address) -> MemoCiphertext {
         let epk = esk.diversified_public(address.diversified_generator());
         let shared_secret = esk
-            .key_agreement_with(&address.transmission_key())
```
Comment (kwf) [inline crypto/src/memo.rs:37]: Unneeded `&`

File: crypto/src/memo.rs [hunk 8]
```diff
@@ -62,7 +62,7 @@ impl MemoPlaintext {
             .key_agreement_with(epk)
             .map_err(|_| anyhow!("could not perform key agreement"))?;
 
-        let key = derive_symmetric_key(&shared_secret, &epk);
```
Comment (kwf) [inline crypto/src/memo.rs:65]: `epk` is already a reference, Rust deferences the `&`

File: crypto/src/memo.rs [hunk 9]
```diff
@@ -105,7 +105,7 @@ mod tests {
 
         let ciphertext = memo.encrypt(&esk, &dest);
 
-        let epk = esk.diversified_public(&dest.diversified_generator());
```
Comment (kwf) [inline crypto/src/memo.rs:108]: Unneeded `&`

File: crypto/src/note.rs [hunk 10]
```diff
@@ -71,7 +71,7 @@ impl Note {
         note_blinding: Fq,
     ) -> Result<Self, Error> {
         Ok(Note {
-            value: value,
```
Comment (kwf) [inline crypto/src/note.rs:74]: Can be abbreviated to `value` since the field name matches

File: crypto/src/note.rs [hunk 11]
```diff
@@ -178,7 +178,7 @@ impl Note {
             .key_agreement_with(epk)
             .map_err(|_| Error::DecryptionError)?;
 
-        let key = derive_symmetric_key(&shared_secret, &epk);
```
Comment (kwf) [inline crypto/src/note.rs:181]: Unneeded `&`

File: crypto/src/note.rs [hunk 12]
```diff
@@ -188,9 +188,9 @@ impl Note {
         let plaintext_bytes: [u8; NOTE_LEN_BYTES] =
             plaintext.try_into().map_err(|_| Error::DecryptionError)?;
 
-        Ok(plaintext_bytes
+        plaintext_bytes
             .try_into()
-            .map_err(|_| Error::DecryptionError)?)
```
Comment (kwf) [inline crypto/src/note.rs:193]: Don't need the `?` if we omit the `Ok`-wrapping

File: crypto/src/note.rs [hunk 13]
```diff
@@ -213,8 +213,8 @@ pub(crate) fn derive_symmetric_key(
     let mut kdf = kdf_params.to_state();
     kdf.update(&shared_secret.0);
     kdf.update(&epk.0);
-    let kdf_output = kdf.finalize();
```
Comment (kwf) [inline crypto/src/note.rs:216]: `let`-binding that can be elided

File: crypto/src/note.rs [hunk 14]
```diff
@@ -238,8 +238,7 @@ impl From<&Note> for [u8; NOTE_LEN_BYTES] {
 
 impl From<&Note> for Vec<u8> {
     fn from(note: &Note) -> Vec<u8> {
-        let mut bytes = Vec::new();
-        bytes.push(NOTE_TYPE);
```
Comment (kwf) [inline crypto/src/note.rs:242]: `Vec::new` followed by `.push(...)` can be turned into a `vec![...]`

File: crypto/src/note.rs [hunk 15]
```diff
@@ -311,9 +310,9 @@ impl Commitment {
     }
 }
 
-impl Into<[u8; 32]> for Commitment {
```
Comment (kwf) [inline crypto/src/note.rs:314]: Replace `Into` impl with `From` impl

File: crypto/src/note.rs [hunk 16]
```diff
@@ -373,7 +372,7 @@ mod tests {
 
         let ciphertext = note.encrypt(&esk);
 
-        let epk = esk.diversified_public(&dest.diversified_generator());
```
Comment (kwf) [inline crypto/src/note.rs:376]: Needless `&`

File: crypto/src/nullifier.rs [hunk 17]
```diff
@@ -14,9 +14,9 @@ pub static NULLIFIER_DOMAIN_SEP: Lazy<Fq> = Lazy::new(|| {
     Fq::from_le_bytes_mod_order(blake2b_simd::blake2b(b"penumbra.nullifier").as_bytes())
 });
 
-impl Into<[u8; 32]> for Nullifier {
```
Comment (kwf) [inline crypto/src/nullifier.rs:17]: Replace `Into` impl with `From` impl

File: crypto/src/proofs/transparent.rs [hunk 18]
```diff
@@ -66,19 +66,15 @@ impl SpendProof {
...
         // Note commitment integrity.
         let s_component_transmission_key = Fq::from_bytes(self.pk_d.0);
-        if s_component_transmission_key.is_err() {
-            proof_verifies = false;
-        } else {
-            let note_commitment_test = note::Commitment::new(
-                self.note_blinding,
-                self.value,
-                self.g_d,
-                s_component_transmission_key.unwrap(),
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:76]: Refactor to avoid this `unwrap` by using `if let` instead

File: crypto/src/proofs/transparent.rs [hunk 19]
```diff
@@ -182,19 +178,15 @@ impl OutputProof {
...
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:192]: Refactor to avoid this `unwrap` by using `if let` instead

File: crypto/src/proofs/transparent.rs [hunk 20]
```diff
@@ -403,9 +395,9 @@ impl TryFrom<transparent_proofs::OutputProof> for OutputProof {
     }
 }
 
-impl Into<Vec<u8>> for SpendProof {
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:406]: Replace `Into` impl with `From` impl

File: crypto/src/proofs/transparent.rs [hunk 21]
```diff
@@ -416,15 +408,15 @@ impl TryFrom<&[u8]> for SpendProof {
     fn try_from(bytes: &[u8]) -> Result<SpendProof, Self::Error> {
         let protobuf_serialized_proof = transparent_proofs::SpendProof::decode(bytes)
             .map_err(|_| ProtoError::ProofMalformed)?;
-        Ok(protobuf_serialized_proof
+        protobuf_serialized_proof
             .try_into()
-            .map_err(|_| ProtoError::ProofMalformed)?)
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:421]: Can avoid `?` by eliding `Ok`-wrapping

File: crypto/src/proofs/transparent.rs [hunk 22]
```diff
@@ -416,15 +408,15 @@ impl TryFrom<&[u8]> for SpendProof {
...
+            .map_err(|_| ProtoError::ProofMalformed)
     }
 }
 
-impl Into<Vec<u8>> for OutputProof {
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:425]: Replace `Into` impl with `From` impl

File: crypto/src/proofs/transparent.rs [hunk 23]
```diff
@@ -435,9 +427,9 @@ impl TryFrom<&[u8]> for OutputProof {
     fn try_from(bytes: &[u8]) -> Result<OutputProof, Self::Error> {
         let protobuf_serialized_proof = transparent_proofs::OutputProof::decode(bytes)
             .map_err(|_| ProtoError::ProofMalformed)?;
-        Ok(protobuf_serialized_proof
+        protobuf_serialized_proof
             .try_into()
-            .map_err(|_| ProtoError::ProofMalformed)?)
```
Comment (kwf) [inline crypto/src/proofs/transparent.rs:440]: Can avoid `?` by eliding `Ok`-wrapping

File: crypto/src/transaction.rs [hunk 24]
```diff
@@ -26,9 +26,9 @@ pub struct TransactionBody {
     pub fee: Fee,
 }
 
-impl Into<Vec<u8>> for TransactionBody {
```
Comment (kwf) [inline crypto/src/transaction.rs:29]: Replace `Into` impl with `From` impl

File: crypto/src/transaction.rs [hunk 25]
```diff
@@ -64,15 +64,9 @@ impl TryFrom<transaction::TransactionBody> for TransactionBody {
             .try_into()
             .map_err(|_| ProtoError::TransactionBodyMalformed)?;
 
-        let expiry_height = proto
-            .expiry_height
-            .try_into()
```
Comment (kwf) [inline crypto/src/transaction.rs:69]: This `try_into` is between the same types, so can be elided

File: crypto/src/transaction.rs [hunk 26]
```diff
@@ -64,15 +64,9 @@ impl TryFrom<transaction::TransactionBody> for TransactionBody {
...
-            .map_err(|_| ProtoError::TransactionBodyMalformed)?;
+        let expiry_height = proto.expiry_height;
 
-        let chain_id = proto
-            .chain_id
-            .try_into()
```
Comment (kwf) [inline crypto/src/transaction.rs:74]: This `try_into` is between the same types, so can be elided

File: crypto/src/transaction/builder.rs [hunk 27]
```diff
@@ -116,8 +116,8 @@ impl Builder {
         let body = output::Body::new(note.clone(), v_blinding, dest, &esk);
         self.value_commitments -= body.value_commitment.0;
 
-        let encrypted_memo = memo.encrypt(&esk, &dest);
```
Comment (kwf) [inline crypto/src/transaction/builder.rs:119]: Needless `&` dereferenced by compiler

File: crypto/src/transaction/builder.rs [hunk 28]
```diff
@@ -116,8 +116,8 @@ impl Builder {
...
-        let ovk_wrapped_key = note.encrypt_key(&esk, &ovk, body.value_commitment);
```
Comment (kwf) [inline crypto/src/transaction/builder.rs:120]: Needless `&` dereferenced by compiler

File: crypto/src/value.rs [hunk 29]
```diff
@@ -57,9 +57,9 @@ impl std::ops::Sub<Commitment> for Commitment {
     }
 }
 
-impl Into<[u8; 32]> for Commitment {
```
Comment (kwf) [inline crypto/src/value.rs:60]: Replace `Into` impl with `From` impl

File: decaf377-fmd/src/clue_key.rs [hunk 30]
```diff
@@ -39,8 +39,8 @@ impl ClueKey {
         // would be nice to avoid this initialization,
         // does collecting into arrays work yet?
         let mut Xs = [Default::default(); MAX_PRECISION];
-        for i in 0..(MAX_PRECISION) {
-            Xs[i] = hkd::derive_public(&root_pub, &root_pub_enc, i as u8);
```
Comment (kwf) [inline decaf377-fmd/src/clue_key.rs:43]: Avoid bounds-check in loop body by using an iterator over pointers to array elements

File: decaf377-fmd/src/detection.rs [hunk 31]
```diff
@@ -39,8 +39,8 @@ impl DetectionKey {
         // would be nice to avoid this initialization,
         // does collecting into arrays work yet?
         let mut xs = [Fr::default(); MAX_PRECISION];
-        for i in 0..MAX_PRECISION {
-            xs[i] = hkd::derive_private(
```
Comment (kwf) [inline decaf377-fmd/src/detection.rs:43]: Avoid bounds-check in loop body by using an iterator over pointers to array elements

File: decaf377-ka/src/lib.rs [hunk 32]
```diff
@@ -72,7 +72,7 @@ impl Secret {
     pub fn key_agreement_with(&self, other: &Public) -> Result<SharedSecret, Error> {
         let pk = decaf377::Encoding(other.0)
             .decompress()
-            .map_err(|_| Error::InvalidPublic(other.clone()))?;
```
Comment (kwf) [inline decaf377-ka/src/lib.rs:75]: Unnecessary `clone` because `other` can instead be moved here

File: decaf377-ka/tests/proptests.rs [hunk 33]
```diff
@@ -1,5 +1,5 @@
 use ark_ff::PrimeField;
-use decaf377;
```
Comment (kwf) [inline decaf377-ka/tests/proptests.rs:2]: Unnecessary root import

File: penumbra/src/app.rs [hunk 34]
```diff
@@ -16,7 +16,7 @@ use penumbra_crypto::{
 };
 use tower_abci::BoxError;
 
-const ABCI_INFO_VERSION: &'static str = env!("VERGEN_GIT_SEMVER");
```
Comment (kwf) [inline penumbra/src/app.rs:19]: Constants are `&'static` by default and don't need to be declared as such

File: penumbra/src/app.rs [hunk 35]
```diff
@@ -202,7 +202,7 @@ impl App {
             self.note_commitment_tree.append(&commitment);
         }
 
-        return true;
```
Comment (kwf) [inline penumbra/src/app.rs:205]: Unnecessary `return`-

File: penumbra/src/app.rs [hunk 36]
```diff
@@ -262,6 +262,6 @@ mod tests {
             .finalize(&mut rng);
 
         // The merkle path is invalid, so this transaction should not verify.
-        assert_eq!(app.verify_transaction(transaction.unwrap()), false);
```
Comment (kwf) [inline penumbra/src/app.rs:265]: `assert_eq!` with a literal boolean can instead be `assert!` with a boolean condition

File: penumbra/src/bin/pcli.rs [hunk 37]
```diff
@@ -117,10 +117,10 @@ async fn main() -> Result<()> {
         }
         Command::Wallet(Wallet::Generate) => {
             if wallet_path.exists() {
-                Err(anyhow::anyhow!(
+                return Err(anyhow::anyhow!(
                     "Wallet path {} already exists, refusing to overwrite it.",
                     wallet_path.display()
-                ))?;
```
Comment (kwf) [inline penumbra/src/bin/pcli.rs:123]: This `?` is always going to throw an error, so we can instead unconditionally `return` the `Err`

File: penumbra/src/bin/pd.rs [hunk 38]
```diff
@@ -55,7 +55,7 @@ async fn main() {
             let _db_insert_dummy_row = db_insert(
                 PenumbraNoteCommitmentTreeAnchor::from(NoteCommitmentTreeAnchor {
                     id: 0,
-                    height: 1337 as i64,
```
Comment (kwf) [inline penumbra/src/bin/pd.rs:58]: Instead of casting, we can write the literal with the appropriate type to begin with

File: proto/build.rs [hunk 39]
```diff
@@ -1,4 +1,3 @@
-use prost_build;
```
Comment (kwf) [inline proto/build.rs:1]: Unnecessary root import

**Review by kwf — COMMENTED** [review]
I've commented next to each of the clippy lint fixes to indicate the problem that was solved in that change.

File: crypto/src/transaction.rs (hunk 26 above)
Comment (redshiftzero) [reply crypto/src/transaction.rs:74]: woops lol

**Review by redshiftzero — APPROVED** [review]
thanks!

**Pull request merged**